[_현] 로또 미션 스터디

nuyhhyun opened the pull request:
- TDD
- enum class
- 일급 컬렉션

nuyhhyun committed:
docs(readMe): readMe 문서 작성

- 로또 게임 정의
- 구현할 기능 목록
- 실행 예시

nuyhhyun committed:
feat(lotto): 로또 구입 금액에 해당하는 로또 수량 계산 기능

docs(readMe): 요구 사항 테마별 들여쓰기로 구분 및 구현한 기능 표시

nuyhhyun committed:
feat(lotto): 로또 구입 금액 1,000원 단위 예외 처리 기능

docs(readMe): 구현한 기능 표시

nuyhhyun committed:
feat(lotto): 로또 구입 금액 양의 정수 예외 처리 기능

docs(readMe): 로또 구입 금액 양의 정수 예외 기능 추가 및 구현한 기능 표시

nuyhhyun committed:
feat(lotto): 로또 구입 금액 양의 정수 예외 처리 기능

docs(readMe): 로또 구입 금액 양의 정수 예외 기능 추가 및 구현한 기능 표시

nuyhhyun committed:
Merge remote-tracking branch 'origin/nuyhhyun' into nuyhhyun

# Conflicts:
#	README.md

nuyhhyun committed:
refactor(lotto): 로또 구입 금액에 따른 갯수 init에 포함

nuyhhyun committed:
feat(lottoTicket): 중복되지 않은 6개 숫자를 갖는 로또 발행 기능

 docs(readMe): 구현한 기능 표시

nuyhhyun committed:
feat(lottoTicket): 로또 범위 1~45 확인 기능

 docs(readMe): 구현한 기능 표시

nuyhhyun committed:
feat(lottoTicket): 로또 오름차순 정렬 기능

 docs(readMe): 구현한 기능 표시

nuyhhyun committed:
feat(winningNumber): 중복되지 않는 당첨 번호 6개와 보너스 번호 발행 기능

 docs(readMe): 구현한 기능 표시

nuyhhyun committed:
feat(winningNumber): 당첨 번호 6개와 보너스 번호 1부터 45 사이 범위 예외 확인 기능

 docs(readMe): 구현한 기능 표시

nuyhhyun committed:
refactor(lottoTicket, winningNumber): 번호 범위와 갯수 상수화

 docs(readMe): UI에 포함되는 요구 사항 삭제

nuyhhyun committed:
feat(lotto): 당첨 여부 확인 기능 구현

 docs(readMe): 로또 번호와 당첨 번호 비교 책임 클래스 로또로 변경, 로또와 순위 요구 사항 추가

nuyhhyun committed:
feat(rank): 당첨 유형 확인 기능 구현

docs(readMe): 구현 기능 표시

nuyhhyun committed:
feat(rank): 당첨 유형마다 횟수 저장 기능 구현

docs(readMe): 로또 컨트롤러단 요구 사항 삭제, 구현한 기능 표시

nuyhhyun committed:
feat(rank): 총 수익 반환 기능 구현

docs(readMe): 수익률 요구 사항 추가, 구현 기능 표시

nuyhhyun committed:
feat(rateOfReturn): 수익률 계산 기능 구현

docs(readMe): 수익률 요구 사항 변경, 구현 기능 표시

nuyhhyun committed:
feat(lotto): 로또 한 장 이상 구매 예외 처리 및 오류 안내 문구 상수화

docs(readMe): 로또 한 장 이상 구매해야 하는 요구 조건 추가, 구현 표시

nuyhhyun committed:
feat(lotto): 로또에서 로또 티켓들을 가지고 있도록 관리하는 기능 구현

docs(readMe): 로또 요구 사항 추가 및 구현 표시

nuyhhyun committed:
refactor(lottoTicket): 일급컬렉션으로써 변수가 하나라는 걸 명확히 하기 위해 변수 선언과 할당 부분 분리

nuyhhyun committed:
feat(UI, controller): 컨트롤러에서 모델 로직들을 이용하여 게임 구현, UI를 위해 입출력 구현

croffle-lover commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,10 @@
+package lotto.view
+
+const val PURCHASE_MONEY_PLEASE = "구입금액을 입력해 주세요."
+class InputView {
+    fun getMoney(): Int {
+        println(PURCHASE_MONEY_PLEASE)
+        val money = readlnOrNull() ?:"0"
+        return money.toInt()
+    }
+}
```
`InputView`가 class여야 하는 이유는 무엇인가요?

또한 `PURCHASE_MONEY_PLEASE`는 `private`이 아닌 이유는 무엇인가요?

nuyhhyun replied:
1. object 사용을 그새 잊어버렸네요..ㅠㅠ 게임을 여러 번 하지 않는 이상 InputView는 한 번만 쓰이니까 object로 싱글톤 사용이 적절해 보입니다!
2. private..를 빠뜨린 것 같습니다.. 특정 값을 담아서 활용하는 게 아니라 상수로 사용하는 것이다 보니 private의 중요성에 더 신경을 못쓴 것 같습니다ㅠ

croffle-lover commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,36 @@
+
+class Lotto(money: Int) {
+    val ticketNumber : Int
+    val lottoTickets : MutableList<List<Int>> = mutableListOf()
```
`Lotto` 클래스는 더 분리 할 수 있어보입니다.

`lottoTickets`를 보았을 때 현재 List<Int>가 로또 숫자 6개를 가지는 역할을 하는 것 같은데요.

`LottoTicket`을 List로 가지는 것이 더 어울려 보입니다.

nuyhhyun replied:
클래스를 자료형으로 갖는 리스트로 바꾸려다 보니 원래 Int형을 자료형으로 가질 때 쓸 수 있는 내장 함수를 어떻게 사용해야 할 지 몰라서 몇 시간을 이것만 붙들고 있었네요ㅠㅠ
잘 해결했는지는 모르겠으나 하나 또 배워갑니다..

croffle-lover commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,36 @@
+    val ticketNumber : Int
+    val lottoTickets : MutableList<List<Int>> = mutableListOf()
+
+    init {
+        require(money != 0) { SHOULD_PURCHASE_AT_LEAST_ONE_TICKET }
+        require(money > 0) { "$money"+IS_NOT_MONEY }
+        require(money % LOTTO_TICKET_PRICE == 0) { "$money"+ IS_NOT_1_000_UNIT}
+
+        ticketNumber = money / LOTTO_TICKET_PRICE
+
+        setLottoTickets()
+    }
```
좋은 검증입니다!

0은 상수화를 안한 이유가 무엇인가요? 틀렸다는게 아니라 의견이 궁금해서 여쭤봅니다!

nuyhhyun replied:
money, LOTTO_TICKET_PRICE 라는 변수명에 '돈'이라는 개념이 있어서 0의 의미가 명확하다 생각했습니다..!
사실 상수화를 여기까지 생각을 못한 것도 있긴 한데 상수화를 하게 된다면 ZERO 정도가 될 것도 같습니다

nuyhhyun replied:
0과 zero가 동일하니 일단 상수화를 시켜둬보겠습니다!

croffle-lover commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,11 @@
+package lotto.model
+
+private val NUMBER_AMOUNT = 6
+private val NUMBER_RANGE = 1..45
+class LottoTicket() {
+    private lateinit var lottoTicket: List<Int>
+    fun makeLottoTicket(): List<Int> {
+        lottoTicket = NUMBER_RANGE.shuffled().chunked(NUMBER_AMOUNT)[0]
+        return lottoTicket.sorted()
+    }
+}
```
로또 번호 자체에도 역할 과 책임이 있어보입니다.

어떤 규칙이 있어야 할까요?

`1 ~ 45의 숫자로 이루어져 있어야한다.`가 있겠죠?

이러한 규칙이 지금은 최상위에 여러 파일에 존재하게 되는데요

`LottoNumber`라는 객체를 만들어서 그 규칙을 가지고 있게끔 하면 어떨까요?

croffle-lover replied:
`LottoTicket`은 6개의 `LottoNumber`로 이루어진 List로 가지고 있음되겠네요!

croffle-lover commented on src/main/kotlin/lotto/model/Rank.kt:
```diff
@@ -0,0 +1,26 @@
+package lotto.model
+
+enum class Rank(val countOfMatch: Int, val winningMoney: Int, val matchBonus:Boolean, var matchAmount: Int) {
+    FIRST(6, 2_000_000_000, false, 0),
+    SECOND(5, 30_000_000, true, 0),
+    THIRD(5, 1_500_000, false, 0),
+    FOURTH(4, 50_000, false, 0),
+    FIFTH(3, 5_000, false, 0),
+    MISS(0, 0, false, 0);
+
+    companion object {
+        fun valueOf(countOfMatch: Int, matchBonus: Boolean): Rank {
+            return values().find {
+                 it.countOfMatch == countOfMatch && it.matchBonus == matchBonus
+            }?: MISS
+        }
+
+        fun saveRank(rank: Rank) {
+            rank.matchAmount += 1
+        }
+
+        fun getWinningMoney(): Int {
+            return values().sumOf { it.winningMoney * it.matchAmount }
+        }
+    }
+}
```
enum class는 가장 완벽한 싱글톤이라는 말이 있습니다.

하지만 지금은 `matchAmount`라는 `var`로 설정되어있어 상태가 가변적입니다.

싱글톤인 객체가 상태를 가지면 안되는 이유는 무엇일까요?

nuyhhyun replied:
싱글톤은 유일하게 사용되기 때문에 var로 상태를 가져서 그 상태가 변경되면 모든 곳에 영향을 미치기 때문인 것 같습니다. 여러 곳에서 상태를 변경하려고 할 경우 값이 꼬일 수도 있을 것도 같습니다.

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/WinningNumberTest.kt:
```diff
@@ -0,0 +1,39 @@
+        assertThat(matchBonus).isNotIn(matchNumber)
+    }
+
+    @Test
+    fun `당첨 번호와 보너스 번호의 범위는 1~45까지이다`() {
+        //given
+        val winning = WinningNumber()
+        val matchNumber = winning.makeMatchNumber()
+        val matchBonus = winning.makeBonusNumber()
+
+        //when
+
+        //then
+        assertThat(matchNumber).allMatch{ it in NUMBER_RANGE }
+        assertThat(matchBonus).isIn(NUMBER_RANGE)
+    }
+}
```
랜덤 값을 어떻게 테스트 한다 했었죠오?!?!?!?!?!?!?!?!?

힌트: `generator`인터페이스

nuyhhyun replied:
이번에도 SAM이 문제인 건지 저번에 랜덤값 하나늘 설정하는 것까진 됐었는데 여러 랜덤값들을 클래스 자료형 리스트로 고정하려니까 방법을 도저히 모르겠습니다.. 이전 것보다 더 시간을 썼는데도 감이 안오네요..ㅠㅠ

croffle-lover reviewed (changes requested):
_현님 테스트 실력이 많이 늘으셨네요!
조그마하게 피드백 몇게 남겨드렸으니 반영해보시고 다시 리뷰 요청 주세요!

nuyhhyun committed:
refactor($InputView): InputView object로 싱글톤화, PURCHASE_MONEY_PLEASE private화

nuyhhyun committed:
refactor($Lotto): init 검증 부분 0을 ZERO_AMOUNT로 상수화

nuyhhyun committed:
refactor($LottoTicketList): LottoTicketList가 List<LottoTicket> 자료형을 갖도록 변경

nuyhhyun committed:
test($LottoListTest): 로또 목록 테스트 생성 및 통과

nuyhhyun committed:
test($LottoNumber): 로또 번호 범위에 맞는 수 발행 클래스와 테스트 생성 및 통과

nuyhhyun committed:
refactor($LottoTicket): 로또 당첨 번호 중복 이슈 해결

nuyhhyun committed:
refactor($Rank): enum class var 상태 가변성 이슈 해결

nuyhhyun committed:
test($numberGenerator): 로또 번호 생성기에 generator 인터페이스 적용

test($lottoNumber): 중복 검사 적용

croffle-lover commented on src/main/kotlin/lotto/model/Rank.kt:
```diff
@@ -0,0 +1,30 @@
+enum class Rank(val countOfMatch: Int, val winningMoney: Int, val matchBonus:Boolean) {
```
[해당글](https://soojin.ro/blog/naming-boolean-variables)을 참고하여 `matchBonus` 변수 명을 변경해볼까요?

croffle-lover commented on src/main/kotlin/lotto/model/RateOfReturn.kt:
```diff
@@ -0,0 +1,10 @@
+
+import kotlin.math.floor
+
+class RateOfReturn {
+    fun getRateOfReturn(money: Int, winningMoney: Int): Double {
+        val rate = winningMoney.toDouble() / money.toDouble()
+        return floor(rate * 100) / 100
+    }
+}
```
`floor`는 내림 입니다! kotlin에서 반올림은 어떻게 할까요?

nuyhhyun replied:
헉 그렇네요

**반올림 : `round`
올림 : `ceil`
내림 : `floor`**

입니다!

croffle-lover commented on src/main/kotlin/lotto/view/ResultView.kt:
```diff
@@ -0,0 +1,55 @@
+private const val RATE_OF_RETURN = "총 수익률은 "
+private const val IS_THIS = "입니다."
+
+class ResultView {
```
`ResultView`도 class일 필요 없을 것 같습니다!

croffle-lover commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,36 @@
+
+class Lotto(money: Int) {
+    val ticketNumber: Int
+    var lottoTicketList = mutableListOf<LottoTicket>()
```
이렇게 되면 `Loto` 인스턴스를 가지고 있는 곳에서 `lottoTicketList`에 아에 엉뚱한 리스트를 집어 넣을 수 있을 것 같습니다.

또한 자료구조에 해당하는 단어는 변수에 사용하지 않는 것이 권장됩니다.
(ex. list, set, map)

croffle-lover commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,36 @@
+private const val SHOULD_PURCHASE_AT_LEAST_ONE_TICKET = "로또는 한 장 이상 구매해야 합니다."
+private const val IS_NOT_MONEY = "은 양의 정수가 아닙니다."
+private const val IS_NOT_1_000_UNIT = "원은 1,000원 단위가 아닙니다."
+private const val ZERO_AMOUNT = 0
```
0은 따로 상수화 안해줘도 
다른사람이 봐도 보편적인 의미를 띈다고 생각합니다.

만약 0과 1이 있을 때 이걸 true false로 생각하고 사용한다면 그 땐 상수화가 필요하겠죠? 

하지만 지금과 같은 경우에는 이 코드를 읽는 사람이 혼동하지 않을 확률이 높기 때문에  ZERO_AMOUNT로 상수화 하지 않아도 될 것 같습니다!

croffle-lover commented on src/main/kotlin/lotto/model/LottoList.kt:
```diff
@@ -0,0 +1,12 @@
+        val numberGenerator = MakeLottoNumber
+        repeat(ticketNumber) {
+            val ticket = LottoTicket()
+                ticket.makeLottoTicket(numberGenerator)
```
항상 바로 `LottoTicket`에 로또티켓을 만들라고함수를 호출해줄거면
`LottoTicket`을 인스턴스화 할때 생성자로 `numberGenerator`를 받고 
내부적으로 init에서 로또티켓을 만드는게 더 좋아보이네요!

croffle-lover commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,25 @@
+
+private val NUMBER_AMOUNT = 6
+
+class LottoTicket() {
+    private lateinit var lottoTicket: List<Int>
+
+    fun makeLottoTicket(numberGenerator: NumberGenerator): LottoTicket {
+        val lottoTicketNumber = mutableListOf<Int>()
+        repeat(NUMBER_AMOUNT) {
+            var number = numberGenerator.generateRandomNumber()
+            while(number in lottoTicketNumber){
+                number = numberGenerator.generateRandomNumber()
+            }
+            lottoTicketNumber.add(number)
+        }
+        lottoTicket = lottoTicketNumber.sorted()
+
+        return this
+    }
+
+    fun getNumbers(): List<Int> {
+        return lottoTicket
+    }
+}
```
`lottoTicket`을 private로 접근제어자를 설정하고 따로 `getNumbers`로 게터로 만들어주면 private 의미가 퇴색됩니다!

nuyhhyun replied:
그것도 그렇네요..
List<Int>의 내장 함수들을 사용하면서 LottoTicket의 형식을 동시에 갖춰두려다 보니 LottoTicket 형식이랑 List<Int> 형식을 번갈아가면서 사용해야 해서 형식 변환을 위해 게터를 설정했는데 이걸 어떻게 해야 할지 잘 모르겠습니다ㅠㅠ
LottoTicket처럼 새로운 형태의 자료형을 만들어도 기존의 자료형을 띄고 있으면 내장 함수를 사용할 수가 있는 건가요? 방법을 이곳저곳에서 찾아봤는데 검색어가 문제인지 찾지를 못했습니다..ㅠㅠ

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/RateOfReturnTest.kt:
```diff
@@ -0,0 +1,22 @@
+import org.junit.jupiter.api.Test
+
+class RateOfReturnTest {
+    @Test
+    fun `구입 금액에 따른 수익률을 구할 수 있다 수익률은 소수점 셋째 자리에서 반올림한다`() {
+        //given
+        val money = 14_000
+        Rank.saveRank(Rank.FIFTH)
+        val winningMoney = Rank.getWinningMoney()
+        val rate = RateOfReturn()
+
+        //when
+        val rateOfReturn = rate.getRateOfReturn(money, winningMoney)
+
+        //then
+        assertThat(rateOfReturn).isEqualTo(0.35)
+    }
+}
```
반올림을 테스트 하려면 셋째 자리가 5이상일 때 올림이 되고 4이하일때 내림이 되는지 테스트하는게 좋을 것 같습니다!

nuyhhyun replied:
테스트 예시 계산해보다 보니 출력 예시가 0.36이어야 하는데 0.35로 돼있더라구요!
ReadME 수정 후에 두 경우로 테스트 다시 작성해보겠습니다!
감사합니다!

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/MakeLottoNumberTest.kt:
```diff
@@ -0,0 +1,28 @@
+
+private val NUMBER_RANGE = 1..45
+
+class MakeLottoNumberTest {
+    @Test
+    fun `로또 번호와 보너스 번호의 범위는 1~45까지이다`() {
+        //given
+        val numberGenerator = MakeLottoNumber
+        val lotto = Lotto(14_000)
+        val match = WinningNumber().makeMatchNumber(numberGenerator).getNumbers()
+        val bonus = WinningNumber().makeBonusNumber(numberGenerator)
+
+        //when
+
+        //then
+        for (lottoTicket in lotto.lottoTicketList)
+            assertThat(lottoTicket.getNumbers()).allMatch { it in NUMBER_RANGE }
+        assertThat(match).allMatch { it in NUMBER_RANGE }
+        assertThat(bonus).isIn( NUMBER_RANGE )
+    }
+}
```
지난번에 말씀 드렸듯 랜덤값이 유효하게 나오냐를 테스트 하는 것은 의미가 없습니다.

사실 잘못 설정해서 10000번 테스트 했는데 운이 좋게 10000번 다 통과하고 10001번째에는 실패할 수 있기 때문이죠.!

nuyhhyun replied:
그렇네요! 범위에 집중하다보니 랜덤값이라는 특성을 간과했습니다😂
ReadME의 기능 요구 사항 모두를 테스트하려다보니 자연스럽게 넣게 되었는데, 랜덤 테스트는 예외라는 사실을 함께 적용을 못시켰네요ㅠㅠ
좀 더 명심해두겠습니다!!

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/LottoTicketTest.kt:
```diff
@@ -0,0 +1,37 @@
+private val NUMBER_AMOUNT = 6
+
+class LottoTicketTest {
+    @Test
+    fun `1개의 로또를 발행할 때 중복되지 않는 6개의 숫자를 무작위로 뽑는다`() {
+        //given
+        val randomGenerator = MakeLottoNumber
+        val lotto = LottoTicket()
+        val lottoTicket = lotto.makeLottoTicket(randomGenerator)
+
+        //when
+        lottoTicket.makeLottoTicket(randomGenerator)
+
+        //then
+        assertThat(lottoTicket.getNumbers().distinct().count()).isEqualTo(NUMBER_AMOUNT)
+    }
```
`distinct()`를 해주는 이유가 무엇일까요?

nuyhhyun replied:
해당 테스트의 목적이 '숫자들이 중복되지 않았는지'여서 중복을 없애는 기능을 하는 `distinct()`를 했을 때 숫자가 `NUMBER_AMOUNT(=6)`이 아니라면 중복값이 있다는 것으로 판별할 수 있을 것 같아 해주었습니다!

croffle-lover reviewed (changes requested):
이번에도 고생하셨습니다. 현님!
피드백 확인해보시고 반영해보세요!

아 그리고! 전체적으로 Lotto 이름이 들어가는 클래스들의 의미와 하는일을 예측하기가 어렵습니다!
다시 한번 네이밍을 고려하심이 어떨까요? 😆

nuyhhyun committed:
refactor($Lotto): 불필요한 0 상수화 제거

nuyhhyun committed:
refactor($ResultView): ResultView를 싱글톤(object)으로 변경

nuyhhyun committed:
refactor($RateOfReturn): 반올림 기능 함수 round로 정정

nuyhhyun committed:
refactor($MakeLottoNumberTest): 랜덤값 테스트 삭제

nuyhhyun committed:
refactor($RateOfReturn): 반올림 기능 함수 수정

- round()는 가장 가까운 수로 변경하는 것이어서 반올림이 제대로 안 될 때가 있다.

test($RateOfReturnTest): 반올림의 세 가지 경우에 대한 테스트로 수정

docs($ReadME): 버림으로 잘못 계산돼있는 예시 반올림으로 계산된 경우로 수정

nuyhhyun committed:
refactor($LottoList): ticket 생성 과정 명료화

- LottoTicket() 인스턴스화와 동시에 makeLottoTicket(numberGenerator) 함수 호출로 가독성 높임

nuyhhyun committed:
refactor($Lotto): `Lotto` 인스턴스를 가진 곳에서 `lottoTicketList`를 수정할 수 없도록 변경

nuyhhyun committed:
refactor($Lotto): 변수명과 클래스명 수정

- matchBonus => matchesBonus
- LottoList => LottoBundle

croffle-lover commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -2,7 +2,7 @@ package lotto.model
 private val NUMBER_AMOUNT = 6
 
-class LottoTicket() {
+class LottoTicket {
```
불필요한 코드 제거 👍

croffle-lover commented on src/main/kotlin/lotto/model/RateOfReturn.kt:
```diff
@@ -1,10 +1,9 @@
+        val rateOfReturn = ((rate + 0.005)*100).toInt() / 100.0
+        return rateOfReturn
     }
 }
```
지금으로썬 굳이 클래스일 필요 없을 것 같습니다!
이 함수가 있을 법한 클래스가 있다면 그곳에 배치 시키는게 나을 것 같네요.
정말 찾을 수 없고 유틸성을 띄는 함수라면 최상위 함수로 위치시키는 것도 방법이겠네요!

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/RateOfReturnTest.kt:
```diff
@@ -2,21 +2,30 @@ package lottoTest.modelTest
-    fun `구입 금액에 따른 수익률을 구할 수 있다 수익률은 소수점 셋째 자리에서 반올림한다`() {
+    @CsvSource("14_000,0.36","16_000,0.31","8_000,0.63")
+    @ParameterizedTest
+    fun `구입 금액에 따른 수익률을 구할 수 있다 수익률은 소수점 셋째 자리에서 반올림한다`(money: Int, expectedRateOfReturn: Double) {
```
CsvSource 활용 bb

croffle-lover commented on src/main/kotlin/lotto/controller/GameManager.kt:
```diff
@@ -0,0 +1,63 @@
+        val match = getMatchNumber(numberGenerator)
+        val bonus = getBonusNumber(numberGenerator)
+        getResult(lotto, match, bonus, money)
+    }
```
`MakeLottoNumber`는 object이다 보니 굳이 변수에 넣어서 관리해줄 필요 없을 것 같아요!

croffle-lover commented on src/main/kotlin/lotto/model/Rank.kt:
```diff
@@ -0,0 +1,32 @@
+    FIFTH(3, 5_000, false),
+    MISS(0, 0, false);
+
+    companion object {
+        private val rankMap = mutableMapOf<Rank, Int>().withDefault { 0 }
+
+        fun valueOf(countOfMatch: Int, matchesBonus: Boolean): Rank {
+            return values().find {
+                it.countOfMatch == countOfMatch && it.matchesBonus == matchesBonus
+            } ?: MISS
+        }
+
+        fun saveRank(rank: Rank) {
+            rankMap[rank] = rankMap.getValue(rank) + 1
+        }
+
+        fun getWinningMoney(): Int {
+            return values().sumOf { it.winningMoney * rankMap.getValue(it) }
+        }
+
+        fun getRank(): MutableMap<Rank, Int> {
+            return rankMap
+        }
+    }
```
싱글톤으로 관리하는 객체가 상태를 가지고 있다면 어떤 위험성이 있을까요?

nuyhhyun replied:
싱글톤은 클래스의 인스턴스를 하나만 생성하여 전역적으로 사용되기 때문에,
- 한 가지의 상태 변화가 해당 싱글톤을 사용하는 모든 곳에 영향을 주어 원치 않는 결과를 가져다주거나
- 다중 스레드 환경일 때 동시에 상태 변화가 일어나지 않아 문제가 생길 수 있고,
- 여러 곳에 의존하기 때문에 독립적인 테스트가 어려워질 수 있다는
등의 위험성이 있을 수 있다고 합니다!

해당 싱글톤이 갖는 mutableMap 때문에 상태를 갖게 되다보니, 싱글톤이 rankMap을 가지고 있지 않게 하는 방식으로 바꾸어보겠습니다!

croffle-lover commented on src/main/kotlin/lotto/model/RateOfReturn.kt:
```diff
@@ -0,0 +1,9 @@
+        val rateOfReturn = ((rate + 0.005)*100).toInt() / 100.0
+        return rateOfReturn
+    }
+}
```
코드 컨벤션이 맞지 않는 것 같아요!
Ctrl + Alt + L을 생활화 하거나 아니면 ktlint를 설정하여 사용하는 것도 방법이겠네요!

croffle-lover commented on src/main/kotlin/lotto/model/WinningNumber.kt:
```diff
@@ -0,0 +1,13 @@
+    fun makeBonusNumber(numberGenerator: NumberGenerator): Int {
+        val bonusNumber = numberGenerator.generateRandomNumber()
+        return bonusNumber
+    }
```
코드 스타일의 영역이겠지만, 해당 코드는 다음과 같이 간추려 질 수 있습니다.

```suggestion
    fun makeBonusNumber(numberGenerator: NumberGenerator): Int = 
        numberGenerator.generateRandomNumber()
```

croffle-lover commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -0,0 +1,10 @@
+        val money = readlnOrNull() ?:"0"
+        return money.toInt()
+    }
+}
```
마찬가지로 코드 포멧팅이 되어있지 않습니다!

croffle-lover commented on src/main/kotlin/lotto/view/ResultView.kt:
```diff
@@ -0,0 +1,55 @@
+            val ticket = lotto.joinToString(", ", "[", "]")
+            println(ticket)
+        }
+    }
```
`lottoTickets`가 컬렉션인 만큼 forEach { } 구문을 써봐도 좋을 것 같습니다!

croffle-lover commented on src/main/kotlin/lotto/view/ResultView.kt:
```diff
@@ -0,0 +1,55 @@
+        println("$ticketNumber" + PURCHASED_LOTTO_THIS_MUCH)
+        for (lottoTicket in lottoTickets) {
+            val lotto = lottoTicket.getNumbers()
+            val ticket = lotto.joinToString(", ", "[", "]")
```
`[`, `]` 은 아마 prefix, suffix에 인자로 들어가는거로 예상이 되는데요!
IDE에선 친절하게 어느 파라미터에 들어갈 것이라는걸 보여주지만, 깃허브와 같은 환경에서는 보이지 않습니다.
때문에 코틀린에서는 `이름있는 인자`라는 걸 지원합니다!

참고: 이펙티브 코틀린 - 아이템 17: 이름 있는 아규먼트를 사용하라
[코틀린 공식문서 - named arguments](https://kotlinlang.org/docs/functions.html#named-arguments)

croffle-lover commented on src/test/kotlin/lottoTest/modelTest/RankTest.kt:
```diff
@@ -0,0 +1,61 @@
+        val rankElse = Rank.valueOf(countOfMatchElse, !matchBonus)
+
+        //then
+        assertAll(
+            { assertThat(rank1).isEqualTo(Rank.FIRST) },
+            { assertThat(rank2).isEqualTo(Rank.SECOND) },
+            { assertThat(rank3).isEqualTo(Rank.THIRD) },
+            { assertThat(rankElse).isEqualTo(Rank.MISS) }
+        )
```
`assertAll`을 지금처럼 사용하는 방법과
여러 라인에 걸쳐 assertThat()을 작성하는 방법의 차이는 무엇일까요?

nuyhhyun replied:
`assertThat()`은 테스트가 실패하면 곧바로 종료되기 때문에, 실패하는 테스트가 처음 생긴다면 해당 실패 테스트 이전의 결과까지만 볼 수 있습니다!
`assertAll()`은 모든 테스트를 실행하기 떄문에, 모든 테스트들에 대한 결과를 볼 수 있어 전체 테스트의 결과를 확인할 수 있습니다.

추가로 `assertSoftly()`는 `assertAll()`처럼 실행되며 그 위치까지 알려주어 테스트가 많을 경우 편리성이 높아집니다!

아래 포스트에 잘 예시와 함께 잘 설명돼 있는 것 같아 첨부합니다!
https://velog.io/@ki960213/assertThat-vs-assertAll-vs-assertSoftly

croffle-lover reviewed (changes requested):
현님 늘 바쁘신데도 꾸준히 하시는군요! 좋습니다!
몇 가지 코멘트 남겼으니 참고하시고 언제든 질문 주셔도 좋습니다!
항상 화이팅입니다!

nuyhhyun committed:
refactor(): 코드 포맷팅, 불필요한 코드 수정, $RateOfReturn 최상위 함수화

nuyhhyun committed:
refactor($Rank): 싱글톤이 상태를 갖지 않도록 수정

no1msh commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -0,0 +1,35 @@
+private const val IS_NOT_MONEY = "은 양의 정수가 아닙니다."
+private const val IS_NOT_1_000_UNIT = "원은 1,000원 단위가 아닙니다."
+
+class Lotto(money: Int) {
+    val ticketNumber: Int
+    val lottoTicketList: List<LottoTicket>
+
+    init {
+        require(money != 0) { SHOULD_PURCHASE_AT_LEAST_ONE_TICKET }
+        require(money > 0) { "$money" + IS_NOT_MONEY }
+        require(money % LOTTO_TICKET_PRICE == 0) { "$money" + IS_NOT_1_000_UNIT }
+
+        ticketNumber = money / LOTTO_TICKET_PRICE
+
+        lottoTicketList = LottoBundle.setLottoTicketBundle(ticketNumber)
+    }
```
money가 0보다 커야한다와 같은 돈 관련 책임이 로또에게 있을까요?
따로 돈 관련 객체를 만들고 그곳에서 검증을 하는게 어떨까요?

no1msh commented on src/main/kotlin/lotto/model/LottoBundle.kt:
```diff
@@ -0,0 +1,13 @@
+package lotto.model
+
+object LottoBundle {
+    fun setLottoTicketBundle(ticketNumber: Int): List<LottoTicket> {
+        val lottoTicketList: MutableList<LottoTicket> = mutableListOf()
+        val numberGenerator = MakeLottoNumber
+        repeat(ticketNumber) {
+            val ticket = LottoTicket().makeLottoTicket(numberGenerator)
+            lottoTicketList.add(ticket)
+        }
+        return lottoTicketList
+    }
+}
```
지난번에도 말씀 드렸지만! List, Set 과 같은 컬렉션에서 쓰이는 단어는 변수명으로 적지 않는 것을 추천드립니다!

nuyhhyun replied:
헉 List 바꾸겠다고 Bundle로 바꿔놓고 정작 변수명을 안바꿨네요ㅋㅋㅋㅠㅠㅜ
인텔리제이에서 Ctrl + Shift + F 로 전체 검색이 되네요..!!
요걸로 다시 바꿔두겠습니다!

no1msh commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,25 @@
+class LottoTicket {
+    private lateinit var lottoTicket: List<Int>
+
+    fun makeLottoTicket(numberGenerator: NumberGenerator): LottoTicket {
+        val lottoTicketNumber = mutableListOf<Int>()
+        repeat(NUMBER_AMOUNT) {
+            var number = numberGenerator.generateRandomNumber()
+            while (number in lottoTicketNumber) {
+                number = numberGenerator.generateRandomNumber()
+            }
+            lottoTicketNumber.add(number)
+        }
+        lottoTicket = lottoTicketNumber.sorted()
+
+        return this
+    }
```
`lottoTicketNumber `는 복수형이 아니라서 변수명만 보고서 컬렉션인줄 인지하지 못합니다!

no1msh commented on src/main/kotlin/lotto/model/LottoTicket.kt:
```diff
@@ -0,0 +1,25 @@
+        return this
+    }
+
+    fun getNumbers(): List<Int> {
+        return lottoTicket
+    }
```
코틀린은 getter를 사용하지 않고 구현이 가능합니다!
아마 lottoTicket을 private로 접근 제한을 하고 노출은 하고싶으셨던거 같은데
backing property에 대해 찾아보고 적용해보시는걸 권장드립니다!

nuyhhyun replied:
backing property 진짜 좋네요ㅠㅠ 계속 이게 문제여서 고민이었었는데 한 큐에 해결했습니다 감사합니다ㅜ

no1msh commented on src/main/kotlin/lotto/model/MakeLottoNumber.kt:
```diff
@@ -0,0 +1,13 @@
+package lotto.model
+
+private val NUMBER_RANGE = 1..45
+
+fun interface NumberGenerator {
+    fun generateRandomNumber(): Int
+}
+
+object MakeLottoNumber : NumberGenerator {
+    override fun generateRandomNumber(): Int {
+        return NUMBER_RANGE.shuffled().first()
+    }
+}
```
같은 파일에 두 객체를 두면 만약 실제 서비스라고 가정할 시 코드의 규모가 커지면 커질수록
파일 분리의 필요성을 느낄 것 같습니다!

no1msh commented on src/main/kotlin/lotto/model/RateOfReturn.kt:
```diff
@@ -0,0 +1,7 @@
+package lotto.model
+
+fun getRateOfReturn(money: Int, winningMoney: Int): Double {
+    val rate = winningMoney.toDouble() / money.toDouble()
+    val rateOfReturn = ((rate + 0.005) * 100).toInt() / 100.0
+    return rateOfReturn
+}
```
수익률을 계산하는 로직은 어떤 객체가 가지고 있는게 좋을까요? 
만약 어울리는게 없다면 새로 만들어 보는 것이 어떨까요?

nuyhhyun replied:
앞 리뷰로 리팩토링 하다가 로또 구입 금액 책임을 갖는 Money라는 객체를 만들었는데, 수익률도 함께 갖고 있으면 어떨까 싶습니다!

no1msh commented on src/main/kotlin/lotto/view/ResultView.kt:
```diff
@@ -0,0 +1,55 @@
+    fun printBonusNumber(bonusNumber: Int) {
+        println(BONUS_NUMBER_PLEASE)
+        println(bonusNumber)
+    }
+
+    fun printWinningResult(rank: Map<Rank, Int>, money: Double) {
+        println(WINNING_RESULT)
+        println(DIVIDE_LINE)
+
+        println(THREE_MATCH + "${rank.getValue(Rank.FIFTH)}" + MATCH_UNIT)
+        println(FOUR_MATCH + "${rank.getValue(Rank.FOURTH)}" + MATCH_UNIT)
+        println(FIVE_MATCH + "${rank.getValue(Rank.THIRD)}" + MATCH_UNIT)
+        println(FIVE_AND_BONUS_MATCH + "${rank.getValue(Rank.SECOND)}" + MATCH_UNIT)
+        println(SIX_MATCH + "${rank.getValue(Rank.FIRST)}" + MATCH_UNIT)
+
+        println(RATE_OF_RETURN + "$money" + IS_THIS)
+    }
```
사실 안드로이드 프레임워크로 들어가면 print 구문을 사용할 경우가 없다시피 하지만, 
사실 입출력은 굉장히 큰 리소스 사용입니다. 
알고리즘 문제를 풀다 보시면 아시게되겠지만, StringBuilder를 사용하지 않고 지금처럼 줄마다 print 구문을 사용하면 상당한 시간 소요를 보실 수 있을 겁니다!

StringBuilder를 사용하여 문자열들을 조합한 뒤 한번에 출력하는 것이 어떨까요?

nuyhhyun replied:
오 그렇군요!! StringBuilder 사용해보니까 더 깔끔하고 보기 좋기까지 한 것 같습니다!
많이 배워갑니다!!

no1msh reviewed (changes requested):
여기까지만 잘 수정해서 마무리 해봅시다!

nuyhhyun committed:
refactor($Money): 로또 구입 금액 조건에 대한 책임 클래스 생성

nuyhhyun committed:
test($Money): 로또 구입 금액 책임 클래스에 대한 테스트 분리

nuyhhyun committed:
refactor($LottoBundle): 변수명에 쓰인 List를 Bundle로 변경

nuyhhyun committed:
refactor($LottoTicket): 컬렉션임을 알 수 있게 하기 위해 변수명lottoTicketNumber를 lottoTicketNumbers로 변경

nuyhhyun committed:
refactor($LottoTicket): backing property 이용하여 getter 제거

nuyhhyun committed:
refactor($RandomNumberGenerator): 난수 생성기와 로또 번호 생성기 코드 분리

nuyhhyun committed:
refactor($RateOfReturn): 수익률 함수 Money 객체로 이동

refactor($Money): Money 객체 object화

nuyhhyun committed:
docs($README.md): 객체와 책임 재정립

nuyhhyun committed:
refactor($ResultView): StringBuilder를 사용해 과도한 print 구문 정리

croffle-lover reviewed (approved):
그동안 정말 고생많으셨습니다. 현님!
앞으로도 훌륭한 개발자가 되기를 바라겠습니다!
